sky/shell/android/update_service_android.cc

```diff
@@ -53,18 +70,87 @@ bool RegisterUpdateService(JNIEnv* env) {
-    : final_path_(base::FilePath(data_dir).AppendASCII("app.skyx")) {
+UpdateTask::UpdateTask(JNIEnv* env,
+                       jobject update_service,
+                       std::string data_dir)
+    : worker_runner_(base::WorkerPool::GetTaskRunner(true)),
+      main_runner_(base::MessageLoop::current()->task_runner()),
+      data_dir_(data_dir) {
   update_service_.Reset(env, update_service);
 
   mojo::ServiceProviderPtr service_provider =
       CreateServiceProvider(Shell::Shared().service_provider_context());
   mojo::ConnectToService(service_provider.get(), &network_service_);
+
+  base::FilePath manifest_path = data_dir_.AppendASCII(kManifestFilename);
+  base::PostTaskAndReplyWithResult(
+      worker_runner_.get(),
+      FROM_HERE,
+      base::Bind(&ReadManifestFromFile, manifest_path),
+      base::Bind(&UpdateTask::OnReadLocalManifest, base::Unretained(this)));
+}
+
+void UpdateTask::Destroy(JNIEnv* env, jobject jcaller) {
+  delete this;
+}
+
+void UpdateTask::OnReadLocalManifest(scoped_ptr<Manifest> manifest) {
+  if (!manifest->IsValid()) {
+    LOG(ERROR) << "Update failed reading local manifest: invalid.";
+    CallOnFinished();
+    return;
+  }
+  current_manifest_ = manifest.Pass();
+  DownloadManifest(current_manifest_->update_url().Resolve(kManifestFilename));
+}
+
+void UpdateTask::DownloadManifest(const GURL& url) {
+  url_loader_ = FetchURL(
+      network_service_.get(), url,
+      base::Bind(&UpdateTask::OnManifestResponse, base::Unretained(this)));
+}
+
+void UpdateTask::OnManifestResponse(mojo::URLResponsePtr response) {
+  mojo::ScopedDataPipeConsumerHandle data;
+  if (response->status_code == 200)
+    data = response->body.Pass();
+  if (!data.is_valid()) {
+    LOG(ERROR) << "Update failed fetching manifest: Server responded "
+               << response->status_code;
+    CallOnFinished();
+    return;
+  }
+
+  base::PostTaskAndReplyWithResult(
+      worker_runner_.get(),
+      FROM_HERE,
+      base::Bind(&ReadManifestFromDataPipe, base::Passed(&data)),
+      base::Bind(&UpdateTask::OnManifestDownloaded, base::Unretained(this)));
 }
 
-void UpdateTask::DownloadAppBundle(const std::string& url) {
-  LOG(INFO) << "Update downloading " << url << " to: " << final_path_.value();
+void UpdateTask::OnManifestDownloaded(scoped_ptr<Manifest> manifest) {
+  if (!manifest->IsValid()) {
+    LOG(ERROR) << "Update failed reading remote manifest: invalid.";
+    CallOnFinished();
+    return;
+  }
+
+  // TODO(mpcomplete): verify it's from the same publisher. (public key check)
+  bool update_available =
+      manifest->version().CompareTo(current_manifest_->version()) > 0;
+  if (!update_available) {
+    LOG(ERROR) << "Update cancelled. No new version.";
+    CallOnFinished();
+    return;
+  }
+
+  // TODO(mpcomplete): replace local manifest with this one.
+  DownloadAppBundle(manifest->update_url().Resolve(kAppBundleFilename));
```

Comment:
How do you know that the app bundle you get is synchronized with the manifest that you got?  What if between the time you fetch the manifest and the time you download the bundle the server updates the bundle to correspond to a different version?